### PR TITLE
Add dto concept

### PR DESCRIPTION
**Pull request by Alexsandr-Yablunovskyi**
DTO and mapper packages are created. New methods (findById and findAllByAuthor) are implemented. EntityNotFoundException is created. All requared configuration are implemented

### CODE REVIEW

File: src/main/java/yablunovskyi/bookstore/BookStoreApplication.java [hunk 1]
```diff
@@ -2,12 +2,17 @@
...
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import yablunovskyi.bookstore.service.BookService;
 
 @SpringBootApplication
 public class BookStoreApplication {
+    private final BookService bookService;
+    
+    public BookStoreApplication(BookService bookService) {
+        this.bookService = bookService;
+    }
```
Comment (validatorr) [inline src/main/java/yablunovskyi/bookstore/BookStoreApplication.java]: I'm not sure that this block of code required here, we are not using it anywhere

File: src/main/java/yablunovskyi/bookstore/controller/BookController.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import yablunovskyi.bookstore.service.BookService;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping(value = "/books")
+public class BookController {
+    private final BookService bookService;
+    
+    @GetMapping
+    public List<BookDto> findAll() {
```
Comment (validatorr) [inline src/main/java/yablunovskyi/bookstore/controller/BookController.java:22]: Correct me if I'm wrong, but we should use find when we want to return Optional(on repository layer) and get on other layers

**Review by validatorr — APPROVED** [review]
Great job!

File: src/main/java/yablunovskyi/bookstore/BookStoreApplication.java (hunk 1 above)
Comment (liliia-ponomarenko) [reply src/main/java/yablunovskyi/bookstore/BookStoreApplication.java]: agree! Let’s  remove it;)

File: src/main/java/yablunovskyi/bookstore/config/MapperConfig.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+import org.mapstruct.InjectionStrategy;
+import org.mapstruct.NullValueCheckStrategy;
+
+@org.mapstruct.MapperConfig(
+        componentModel = "spring",
+        injectionStrategy = InjectionStrategy.CONSTRUCTOR,
+        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
+        implementationPackage = "<PACKAGE_NAME>.impl"
+)
+public class MapperConfig {
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/config/MapperConfig.java]: it is better to specify @MapperConfig directly on mapper class

File: src/main/java/yablunovskyi/bookstore/controller/BookController.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        return bookService.findAll();
+    }
+    
+    @GetMapping("/{id}")
+    public BookDto findById(@PathVariable Long id) {
+        return bookService.findById(id);
+    }
+    
+    @GetMapping("/by-author")
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/controller/BookController.java]: you will realize `/search` in the next task ;)

File: src/main/java/yablunovskyi/bookstore/controller/BookController.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    public List<BookDto> findAllByAuthor(@RequestParam String author) {
+        return bookService.findAllByAuthor(author);
+    }
+    
+    @PostMapping
+    public BookDto save(@RequestBody CreateBookRequestDto requestDto) {
+        return bookService.save(requestDto);
+    }
+
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/controller/BookController.java]:
```suggestion
```

File: src/main/java/yablunovskyi/bookstore/dto/BookDto.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package yablunovskyi.bookstore.dto;
+
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Data
+public class BookDto {
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/dto/BookDto.java]: it is better to use records for DTO (apply  to all dtos) 

upgrade your maven-checkstyle-plugin to 3.3.0 if Checkstyle fires at records

File: src/main/java/yablunovskyi/bookstore/model/Book.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Entity
+@Table(name = "books")
+@Data
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/model/Book.java:14]: Lombok is a great library but be [attentive when using it with Entity](https://jpa-buddy.com/blog/lombok-and-jpa-what-may-go-wrong/) JFYI

File: src/main/java/yablunovskyi/bookstore/model/Book.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    
+    @Column(nullable = false, unique = true)
+    private String isbn;
+    
+    @Column(nullable = false)
+    private BigDecimal price;
+    
+    private String description;
+    
+    private String coverImage;
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/model/Book.java]:
```suggestion
    private BigDecimal price;
    private String description;
    private String coverImage;
```

File: src/main/java/yablunovskyi/bookstore/repository/impl/BookRepositoryImpl.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+            throw new DataProcessingException("Can't creat a new book", e);
+        }
+    }
+    
+    @Override
+    public Optional<Book> findById(Long id) {
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            Book book = entityManager.find(Book.class, id);
+            return Optional.ofNullable(book);
+        }
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/repository/impl/BookRepositoryImpl.java]: Let’s  add catch block

File: src/main/java/yablunovskyi/bookstore/repository/impl/BookRepositoryImpl.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    
+    @Override
+    public List<Book> findAllByAuthor(String author) {
+        String lowerCaseAuthor = author.toLowerCase();
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            return entityManager.createQuery(
+                    "SELECT b FROM Book b WHERE lower(b.author) LIKE :author", Book.class)
+                    .setParameter("author", "%" + lowerCaseAuthor + "%")
+                    .getResultList();
+
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/repository/impl/BookRepositoryImpl.java]:
```suggestion
```

File: src/main/java/yablunovskyi/bookstore/repository/impl/BookRepositoryImpl.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    @Override
+    public Optional<Book> findById(Long id) {
+        try (EntityManager entityManager = entityManagerFactory.createEntityManager()) {
+            Book book = entityManager.find(Book.class, id);
+            return Optional.ofNullable(book);
+        }
+    }
+    
+    @Override
+    public List<Book> findAllByAuthor(String author) {
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/repository/impl/BookRepositoryImpl.java]: You can leave it for now but don’t forget remove it when you add `search `

File: src/main/java/yablunovskyi/bookstore/service/impl/BookServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    
+    @Override
+    public BookDto findById(Long id) {
+        Book book = bookRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Can't find book by id: %d".formatted(id))
+        );
+        /*Optional<Book> optionalBook = bookRepository.findById(id);
+        if (optionalBook.isPresent()) {
+            return bookMapper.toDto(optionalBook.get());
+        }*/
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/service/impl/BookServiceImpl.java]: Let’s remove commented code ;)

File: src/main/java/yablunovskyi/bookstore/service/impl/BookServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+        return bookMapper.toDto(book);
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/service/impl/BookServiceImpl.java]: there is a map method, so that you can do everything in a single chain

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  improve your solution ;)

File: src/main/java/yablunovskyi/bookstore/controller/BookController.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    @GetMapping("/{id}")
+    public BookDto findById(@PathVariable Long id) {
+        return bookService.findById(id);
+    }
+    
+    /* @GetMapping("/by-author")
+    public List<BookDto> findAllByAuthor(@RequestParam String author) {
+        return bookService.findAllByAuthor(author);
+    }
+    */
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/controller/BookController.java:35]: Better don’t  leave commented code

File: src/main/java/yablunovskyi/bookstore/service/BookService.java [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+import java.util.List;
+import yablunovskyi.bookstore.dto.BookDto;
+import yablunovskyi.bookstore.dto.CreateBookRequestDto;
+
+public interface BookService {
+    BookDto save(CreateBookRequestDto requestDto);
+    
+    BookDto findById(Long id);
+    
+    /*List<BookDto> findAllByAuthor(String author);*/
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/service/BookService.java:12]: the same

File: src/main/java/yablunovskyi/bookstore/service/impl/BookServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+                .map(bookMapper::toDto)
+                .toList();
+    }
+    
+    /*@Override
+    public List<BookDto> findAllByAuthor(String author) {
+        return bookRepository.findAllByAuthor(author).stream()
+                .map(bookMapper::toDto)
+                .toList();
+    }*/
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/service/impl/BookServiceImpl.java:45]: the same

**Review by liliia-ponomarenko — APPROVED** [review]
Great! But fix my comment before merging ;)

**Pull request merged**